mysql/mysqldump.go

```diff
@@ -0,0 +1,585 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mysql
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/cloudspannerecosystem/harbourbridge/internal"
+	"github.com/cloudspannerecosystem/harbourbridge/schema"
+	"github.com/pingcap/parser"
+	"github.com/pingcap/parser/ast"
+	driver "github.com/pingcap/tidb/types/parser_driver"
+)
+
+// TODO: consider refactoring mysqldump.go and pgdump.go to extract
+// common code-paths/logic into a shared set of utils.
+
+var valuesRegexp = regexp.MustCompile("\\((.*?)\\)")
+var insertRegexp = regexp.MustCompile("INSERT\\sINTO\\s(.*?)\\sVALUES\\s")
+
+// MysqlSpatialDataTypes is an array of all mysql spatial data types.
+var MysqlSpatialDataTypes = []string{"geometrycollection", "multipoint", "multilinestring", "multipolygon", "point", "linestring", "polygon", "geometry"}
+
+// ProcessMySQLDump reads mysqldump data from r and does schema or data conversion,
+// depending on whether conv is configured for schema mode or data mode.
+// In schema mode, ProcessMySQLDump incrementally builds a schema (updating conv).
+// In data mode, ProcessMySQLDump uses this schema to convert MySQL data
+// and writes it to Spanner, using the data sink specified in conv.
+func ProcessMySQLDump(conv *internal.Conv, r *internal.Reader) error {
+	for {
+		startLine := r.LineNumber
+		startOffset := r.Offset
+		b, stmts, err := readAndParseChunk(conv, r)
+		if err != nil {
+			return err
+		}
+		for _, stmt := range stmts {
+			isInsert := processStatement(conv, stmt)
+			internal.VerbosePrintf("Parsed SQL command at line=%d/fpos=%d: %d stmts (%d lines, %d bytes) Insert Statement=%v\n", startLine, startOffset, 1, r.LineNumber-startLine, len(b), isInsert)
+		}
+		if r.EOF {
+			break
+		}
+	}
+	if conv.SchemaMode() {
+		schemaToDDL(conv)
+		conv.AddPrimaryKeys()
+	}
+	return nil
+}
+
+// readAndParseChunk parses a chunk of mysqldump data, returning the bytes read,
+// the parsed AST (nil if nothing read), error and whether we've hit end-of-file.
```

Comment:
How about adding an explanation here about how there is a limit on the size of insert statements. Suggest:

// In effect, we proceed through the file, statement by statement. Many
// statements (e.g. DDL statements) are small, but insert statements can
// be large. Fortunately mysqldump limits the size of insert statements
// (default is 24MB, but configurable via --max-allowed-packet), and so
// the chunks of file we read/parse are manageable, even for mysqldump
// files containing tens or hundreds of GB of data.

Reply:
Sure. Done.